File: Samples/Logging/cpp/FileLoggingSessionScenario.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
//
//*********************************************************

#include "pch.h"
#include "FileLoggingSessionScenario.h"
#include "LoggingScenarioHelpers.h"

#define SCENARIO_PREFIX                     L"FileLoggingSessionScenario_"
#define DEFAULT_SESSION_NAME                SCENARIO_PREFIX L"Session"
#define DEFAULT_CHANNEL_NAME                SCENARIO_PREFIX L"Channel"
#define OUR_SAMPLE_APP_LOG_FILE_FOLDER_NAME SCENARIO_PREFIX L"LogFiles"
#define LOGGING_ENABLED_SETTING_KEY_NAME    SCENARIO_PREFIX L"LoggingEnabled"
#define LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME SCENARIO_PREFIX L"LogFileGeneratedBeforeSuspend"

using namespace concurrency;
using namespace Windows::Foundation::Diagnostics;
using namespace Windows::Storage;
using namespace SDKTemplate;

FileLoggingSessionScenario^ FileLoggingSessionScenario::_instance;

FileLoggingSessionScenario::FileLoggingSessionScenario()
{
    _logFileGeneratedCount = 0;
    _isPreparingForSuspend = false;
    _isBusy = false;

    // Create the logging channel.
    // When an app logs messages to a channel, the messges will go 
    // to any sessions which are consuming messages from the channel.
#pragma warning(push)
#pragma warning(disable : 4973)
    _channel = ref new Windows::Foundation::Diagnostics::LoggingChannel(DEFAULT_CHANNEL_NAME);
#pragma warning(pop)

    // This sample listens to the channel's events so it can track 
    // information about the channel's enablement status. 
    _channel->LoggingEnabled += ref new Windows::Foundation::TypedEventHandler<Windows::Foundation::Diagnostics::ILoggingChannel ^, Platform::Object ^>(this, &FileLoggingSessionScenario::OnChannelLoggingEnabled);

    App::Current->Suspending += ref new Windows::UI::Xaml::SuspendingEventHandler(this, &FileLoggingSessionScenario::OnAppSuspending);
    App::Current->Resuming += ref new Windows::Foundation::EventHandler<Platform::Object ^>(this, &FileLoggingSessionScenario::OnAppResuming);

    // If the app is being launched (not resumed), the 
    // following call will activate logging if it had been 
    // active at the time of the last suspend. 
    ResumeLoggingIfApplicable();
}

FileLoggingSessionScenario::~FileLoggingSessionScenario()
{
}

void FileLoggingSessionScenario::OnAppSuspending(Platform::Object ^sender, Windows::ApplicationModel::SuspendingEventArgs ^e)
{
    (void) sender;    // Unused parameter

    // Get a deferral before performing any async operations
    // to avoid suspension prior to LoggingScenario completing 
    // PrepareToSuspendAsync().
    auto deferral = e->SuspendingOperation->GetDeferral();
    PrepareToSuspendAsync().then([=](task<void> previousTask)
    {
        // From LoggingScenario's perspective, it's now okay to 
        // suspend, so release the deferral. 
        deferral->Complete();
    });
}

void FileLoggingSessionScenario::OnAppResuming(Platform::Object ^sender, Platform::Object ^args)
{
    // If logging was active at the last suspend,
    // ResumeLoggingIfApplicable will re-activate 
    // logging.
    ResumeLoggingIfApplicable();
}

void FileLoggingSessionScenario::OnChannelLoggingEnabled(ILoggingChannel^ sender, Platform::Object^ args)
{
    // This method is called when the channel is informing us of channel-related state changes.
    // Save new channel state. These values can be used for advanced logging scenarios where, 
    // for example, it's desired to skip blocks of logging code if the channel is not being
    // consumed by any sessions. 
    _isChannelEnabled = sender->Enabled;
    _channelLoggingLevel = sender->Level;
}

// When the log file gets large, the system closes it and starts a new one.
// The LogFileGenerated event is raised to give the app a chance to move
// the recently-generated log file to a safe place. When the handler returns,
// the FileLoggingSession may reuse the file name for a new log file.
// This function moves the log file to an app-defined location.
void FileLoggingSessionScenario::OnLogFileGenerated(IFileLoggingSession^ sender, LogFileGeneratedEventArgs^ args)
{
    InterlockedIncrement(&_logFileGeneratedCount);

    // ContinuationData allows information to be
    // shared across the below async operations.
    struct ContinuationData
    {
        StorageFolder^ ourSampleAppLogFolder;
        StorageFile^ generatedLogFile;
        Platform::String^ logFileName;
    };
    auto data = std::make_shared<ContinuationData>();
    data->generatedLogFile = args->File;
    // Create the app-defined folder under the app's local folder.
    // An app defines where/how it wants to store log files. 
    // This sample uses OUR_SAMPLE_APP_LOG_FILE_FOLDER_NAME under
    // the app's local folder. 
    create_task(ApplicationData::Current->LocalFolder->CreateFolderAsync(OUR_SAMPLE_APP_LOG_FILE_FOLDER_NAME, CreationCollisionOption::OpenIfExists))
        .then([this, data](StorageFolder^ ourSampleAppLogFolder)
        {
            // After creating/opening our sample app's log folder, 
            // save it for use in a continuation further below. 
            data->ourSampleAppLogFolder = ourSampleAppLogFolder;
            // Create a new log file name, and save it for use 
            // in a continuation further below. 
            data->logFileName = "Log-" + ref new Platform::String(GetTimeStamp().c_str()) + ".etl";
            // Move the log file to our sample app's log folder. 
            return data->generatedLogFile->MoveAsync(ourSampleAppLogFolder, data->logFileName, Windows::Storage::NameCollisionOption::FailIfExists);
        })
        .then([this, data](task<void> previousTask)
        {
            // Update the UI to show the new log file.
            if (_isPreparingForSuspend == false)
            {
                StatusChanged(this, ref new LoggingScenarioEventArgs(LoggingScenarioEventType::LogFileGenerated, std::wstring((data->ourSampleAppLogFolder->Path + "\\" + data->logFileName)->Data())));
            }
        });
}

task<StorageFile^> FileLoggingSessionScenario::CloseSessionSaveFinalLogFile()
{

    // ContinuationData allows information to be
    // shared across the below async operations.
    struct ContinuationData
    {
        StorageFolder^ ourSampleAppLogFolder;
        Platform::String^ logFileName;
    };
    auto data = std::make_shared<ContinuationData>();

    // Tell the session to save its final log file.
    return create_task(_session->CloseAndSaveToFileAsync())
        .then([this, data](StorageFile^ finalFileBeforeSuspend)
        {
            if (finalFileBeforeSuspend != nullptr)
            {
                // A final log file was created. 
                // Move it to the sample-defined log folder. 
                // First, get create/open our sample-defined log folder.
                return create_task(ApplicationData::Current->LocalFolder->CreateFolderAsync(OUR_SAMPLE_APP_LOG_FILE_FOLDER_NAME, CreationCollisionOption::OpenIfExists))
                    .then([this, finalFileBeforeSuspend, data](StorageFolder^ ourSampleAppLogFolder)
                    {
                        // Next, move the log file to the sample-defined log folder. 
                        data->ourSampleAppLogFolder = ourSampleAppLogFolder;
                        data->logFileName = "Log-" + ref new Platform::String(GetTimeStamp().c_str()) + ".etl";
                        return finalFileBeforeSuspend->MoveAsync(ourSampleAppLogFolder, data->logFileName, Windows::Storage::NameCollisionOption::FailIfExists);
                    })
                    .then([this, data](task<void> previousTask)
                    {
                        // Return a StorageFile representing the log file. 
                        return StorageFile::GetFileFromPathAsync(data->ourSampleAppLogFolder->Path + "\\" + data->logFileName);
                    });
            }
            
            // Async operation need to return consistent types.
            // If the session did not produce any log file, 
            // return a task which produces a null StorageFile.
            return task<StorageFile^>([]() -> StorageFile^
            {
                return nullptr;
            });     
        });
}

// Handle a request to toggle logging enable/disable status.
task<bool> FileLoggingSessionScenario::ToggleLoggingEnabledDisabledAsync()
{
    IsBusy = true;
    try
    {
        if (_session != nullptr)
        {
            // Save any final log file.
            return create_task(CloseSessionSaveFinalLogFile()).then([this](StorageFile^ finalLogFilePath)
            {
                // Cleanup the session and let the sample UI know of the new status.
                delete _session;
                _session = nullptr;
                StatusChanged(this, ref new LoggingScenarioEventArgs(LoggingScenarioEventType::LogFileGeneratedAtDisable, finalLogFilePath == nullptr ? L"" : finalLogFilePath->Path->Data()));
                SetAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME, false);
                StatusChanged(this, ref new LoggingScenarioEventArgs(false));
                IsBusy = false;
                return false;
            });
        }
        else
        {
            // Enable logging and let the UI know of the new status.
            StartLogging();
            SetAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME, true);
            StatusChanged(this, ref new LoggingScenarioEventArgs(true));
            IsBusy = false;
            return task<bool>([]() -> bool
            {
                return true;
            });
        }
    }
    catch (...)
    {
        IsBusy = false;
        throw;
    }
}

void FileLoggingSessionScenario::StartLogging()
{
    // If no session exists, create one.
    // NOTE: There are use cases where an application
    // may want to create only a channel for sessions outside
    // of the application itself. See MSDN for details. This
    // sample is the common scenario of an app logging events
    // which it wants to place in its own log file, so it creates
    // a session and channel as a pair. The channel is created 
    // during construction of this LoggingScenario class so 
    // it already exsits by the time this function is called. 
    if (_session == nullptr)
    {
        _session = ref new Windows::Foundation::Diagnostics::FileLoggingSession(DEFAULT_SESSION_NAME);
        _session->LogFileGenerated += ref new Windows::Foundation::TypedEventHandler<IFileLoggingSession ^, LogFileGeneratedEventArgs ^>(this, &FileLoggingSessionScenario::OnLogFileGenerated);
    }

    // This sample adds the channel at level "warning" to 
    // demonstrated how messages logged at more verbose levels
    // are ignored by the session. 
    _session->AddLoggingChannel(_channel, LoggingLevel::Warning);
}

// Prepare this scenario for suspend. 
task<void> FileLoggingSessionScenario::PrepareToSuspendAsync()
{
    if (_session != nullptr)
    {
        _isPreparingForSuspend = true;
        // Before suspend, save any final log file.
        return create_task(CloseSessionSaveFinalLogFile()).then([this](StorageFile^ finalLogFile)
        {
            // Cleanup the session.
            delete _session;
            _session = nullptr;
            // Save values used when the app is resumed or started later.
            // Logging is enabled.
            SetAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME, true);
            // If there's a final log file...
            if (finalLogFile != nullptr)
            {
                // Save the log file name saved at suspend so the sample UI can be 
                // updated on resume with that information. 
                SetAppLocalSettingsValue(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME, finalLogFile->Path);
            }
            else
            {
                SetAppLocalSettingsValue(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME, nullptr);
            }
            _isPreparingForSuspend = false;
            return task<void>([]() -> void {});
        });
    }
    else
    {
        // Save values used when the app is resumed or started later.
        // Logging is not enabled and no log file was saved.
        SetAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME, false);
        SetAppLocalSettingsValue(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME, nullptr);
        return task<void>([]() -> void {});
    }
}

// This is called when the app is either resuming or starting. 
// It will enable logging if the app has never been started before
// or if logging had been enabled the last time the app was running.
void FileLoggingSessionScenario::ResumeLoggingIfApplicable()
{
    bool loggingEnabled;
    if (IsAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME))
    {
        loggingEnabled = GetAppLocalSettingsValueAsBool(LOGGING_ENABLED_SETTING_KEY_NAME);
    }
    else
    {         
        SetAppLocalSettingsValue(LOGGING_ENABLED_SETTING_KEY_NAME, true);
        loggingEnabled = true;
    }

    if (loggingEnabled)
    {
        StartLogging();
    }

    // When the sample suspends, it retains state as to whether or not it had
    // generated a new log file at the last suspension. This allows any
    // UI to be updated on resume to reflect that fact. 
    if (IsAppLocalSettingsValue(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME))
    {
        Platform::String^ pathOfLogFileGeneratedBeforeSuspend = GetAppLocalSettingsValueAsString(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME);
        if (pathOfLogFileGeneratedBeforeSuspend != nullptr)
        {
            StatusChanged(this, ref new LoggingScenarioEventArgs(LoggingScenarioEventType::LogFileGeneratedAtSuspend, pathOfLogFileGeneratedBeforeSuspend->Data()));
            SetAppLocalSettingsValue(LOGFILEGEN_BEFORE_SUSPEND_SETTING_KEY_NAME, nullptr);
        }
    }
}

// DoScenarioAsync creates a worker thread to simulate application activity.
// In actual use, logging functions would be called from various parts of a 
// larger application. The purpose of the below concise function is to show
// how logging works over periods of time when lots of messages are logged.
// Primarily, it shows how log files are generated when each log file reaches
// its maximum capacity. This function logs information in a loop 
// that exits after 3 log files have been generated. After each log file is 
// generated, the sample's UI is updated with the log file's name.
task<void> FileLoggingSessionScenario::DoScenarioAsync()
{
    // Tell the UI the scenario is busy working.
    IsBusy = true;
    try
    {
        // ContinuationData allows information to be
        // shared across the below async operations.
        struct ContinuationData
        {
            int messageIndex;
            int startFileCount;
            std::function<concurrency::task<void>(const std::shared_ptr<ContinuationData>&)> func;
        };

        // Create an instance of ContinuationData to share
        // across the below async tasks.
        auto data = std::make_shared<ContinuationData>();

        // A message index counter which increments for each
        // logging call. 
        data->messageIndex = 0;

        // Get the count of log files created so far. 
        // The code below will execute until 3 more 
        // log files are created.
        data->startFileCount = LogFileGeneratedCount;

        // This is the main worker function. It is 
        // called asynchronously until 3 new log files
        // have been created.
        data->func = [this](const std::shared_ptr<ContinuationData>& data)
        {
            const int NUMBER_OF_LOG_FILES_TO_GENERATE = 3;

            // To demo logging, loop and log messages until 3 log files have been produced.
            // Each time a new log is created, the FileLoggingSession will notify this sample
            // via the FileLoggingSession::LogFileGenerated event. 
            // While not 3 files yet, execute the following code...
            while (LogFileGeneratedCount - data->startFileCount < NUMBER_OF_LOG_FILES_TO_GENERATE)
            {
                // Since the channel is added to the session at level Warning, the following 
                // message is logged because it is logged at level LoggingLevel.Critical.
                Platform::String^ messageToLog =
                    "Message=" +
                    (++data->messageIndex).ToString() +
                    ": Lorem ipsum dolor sit amet, consectetur adipiscing elit.In ligula nisi, vehicula nec eleifend vel, rutrum non dolor.Vestibulum ante ipsum " +
                    "primis in faucibus orci luctus et ultrices posuere cubilia Curae; Curabitur elementum scelerisque accumsan. In hac habitasse platea dictumst.";
                _channel->LogMessage(messageToLog, LoggingLevel::Critical);

                // Since the channel is added to the session at level Warning, the following 
                // message is *not* logged because it is logged at LoggingLevel.Information.
                messageToLog =
                    "Message=" +
                    (++data->messageIndex).ToString() +
                    ": Lorem ipsum dolor sit amet, consectetur adipiscing elit.In ligula nisi, vehicula nec eleifend vel, rutrum non dolor.Vestibulum ante ipsum " +
                    "primis in faucibus orci luctus et ultrices posuere cubilia Curae; Curabitur elementum scelerisque accumsan. In hac habitasse platea dictumst.";
                this->_channel->LogMessage(messageToLog, LoggingLevel::Information);

                int value = 1000000; // one million, 7 digits, 4-bytes as an int, 14 bytes as a wide character string.
                _channel->LogMessage("Value #" + (++data->messageIndex).ToString() + "  " + value.ToString(), LoggingLevel::Critical); // value is logged as 14 byte wide character string.
                _channel->LogValuePair("Value #" + (++data->messageIndex).ToString(), value, LoggingLevel::Critical); // value is logged as a 4-byte integer.

                //
                // Pause every once in a while to simulate application activity outside of logging.
                //

                if (data->messageIndex % 50 == 0)
                {
                    // Create a task which calls this function again in 5 msecs.
                    // When this function is called again, it will continue its work.
                    return create_delayed_task(std::chrono::milliseconds(5), [=] { return data->func(data); });
                }
            }

            // 3 log files created, nothing more to do, stop.
            return task<void>([]() -> void {});
        };

        // Create the first task of several which will create 3 log files, 
        // and then execute the following continuation code when finished...
        return data->func(data).then([=](task<void> previousTask)
        {
            // This executes after 3 new log files have been created or
            // in the event of an unhandled exception in the above code.
            try
            {
                // Tell the sample UI that the scenario is not busy any longer.
                IsBusy = false;

                // The following 'get' will re-throw any exceptions which 
                // occurred during the above code (data->func above).
                previousTask.get(); 
            }
            catch (...)
            {
                OutputDebugString(L"Exception during DoScenario.");
                throw;
            }
        });
    }
    catch (...)
    {
        IsBusy = false;
        throw;
    }
}

File: Samples/Logging/js/js/fileLoggingSessionScenario.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    function getTimeStamp() {
        var d = new Date();
        var ts = String("0" + d.getYear()).slice(-2) +
                 String("0" + d.getMonth()).slice(-2) +
                 String("0" + d.getDay()).slice(-2) +
                 "-" +
                 String("0" + d.getHours()).slice(-2) +
                 String("0" + d.getMinutes()).slice(-2) +
                 String("0" + d.getSeconds()).slice(-2) +
                 String("00" + d.getMilliseconds()).slice(-3);
        return ts;
    }

    var FileLoggingSessionScenario = WinJS.Class.define(
        function () {
            this._channel = new Windows.Foundation.Diagnostics.LoggingChannel("FileLoggingSession_Channel");
            this._channel.addEventListener("loggingenabled", this._onLoggingEnabled);

            Windows.UI.WebUI.WebUIApplication.addEventListener("suspending", this._onAppSuspension);
            Windows.UI.WebUI.WebUIApplication.addEventListener("resuming", this._onAppResume);
        },
        {
            doScenarioAsync: function () {
                this._setBusyStatus(true);
                try {

                    var data = { scenario: this, messageIndex: 0, startFileCount: this._logFileGeneratedCount };

                    function logMessagesAndPauseAsync(data) {

                        var totalFilesCreatedSoFar = data.scenario._logFileGeneratedCount - data.startFileCount;
                        if (totalFilesCreatedSoFar >= 3) {
                            // At least 3 files created, stop.
                            return new WinJS.Promise(function (completionFunction, errorFunction, progressFunction) {
                                data.scenario._setBusyStatus(false);
                                completionFunction();
                            });
                        } else {

                            var channel = data.scenario._channel;

                            for (var index = 0; index < 50; index++) {

                                // Since the channel is added to the session at level Warning,
                                // the following is logged because it is logged at level LoggingLevel.Critical.
                                var messageToLog =
                                    "Message=" +
                                    (++data.messageIndex) +
                                    ": Lorem ipsum dolor sit amet, consectetur adipiscing elit.In ligula nisi, vehicula nec eleifend vel, rutrum non dolor.Vestibulum ante ipsum " +
                                    "primis in faucibus orci luctus et ultrices posuere cubilia Curae; Curabitur elementum scelerisque accumsan. In hac habitasse platea dictumst.";
                                channel.logMessage(messageToLog, Windows.Foundation.Diagnostics.LoggingLevel.critical);

                                // Since the channel is added to the session at level Warning,
                                // the following is *not* logged because it is logged at LoggingLevel.Information.
                                messageToLog =
                                    "Message=" +
                                    (++data.messageIndex) +
                                    ": Lorem ipsum dolor sit amet, consectetur adipiscing elit.In ligula nisi, vehicula nec eleifend vel, rutrum non dolor.Vestibulum ante ipsum " +
                                    "primis in faucibus orci luctus et ultrices posuere cubilia Curae; Curabitur elementum scelerisque accumsan. In hac habitasse platea dictumst.";
                                channel.logMessage(messageToLog, Windows.Foundation.Diagnostics.LoggingLevel.information);

                                var value = 1000000; // 1 million, 7 digits, 4-bytes as an integer, 14 bytes as a wide character string.
                                channel.logMessage("Value #" + (++data.messageIndex).toString() + "  " + value, Windows.Foundation.Diagnostics.LoggingLevel.critical); // 'value' is logged as 14 byte wide character string.
                                channel.logValuePair("Value #" + (++data.messageIndex).toString(), value, Windows.Foundation.Diagnostics.LoggingLevel.critical); // 'value' is logged as a 4-byte integer.
                            }

                            // Pause every once in a while to simulate application
                            // activity outside of logging.
                            return WinJS.Promise.timeout(25).then(function () {

                                return logMessagesAndPauseAsync(data);
                            });
                        }
                    }

                    // Log messages until at least 3 log files are created.
                    return logMessagesAndPauseAsync(data);

                } catch (e) {
                    this._setBusyStatus(false);
                    throw e;
                }
            },
            toggleLoggingEnabledDisabledAsync: function () {
                this._setBusyStatus(true);
                try {
                    if (this._session !== null) {
                        var data = { scenario: this };
                        return this._closeSessionSaveFinalLogFile().then(function (finalLogFilePath) {
                            data.scenario._session.close();
                            data.scenario._session = null;
                            data.scenario._dispatchStatusChanged({ type: "LogFileGeneratedAtDisable", logFilePath: finalLogFilePath });
                            Windows.Storage.ApplicationData.current.localSettings.values["FileLoggingSession_LoggingEnabled"] = false;
                            data.scenario._dispatchStatusChanged({ type: "LoggingEnabledDisabled", enabled: false });
                            data.scenario._setBusyStatus(false);
                            return false;
                        });
                    } else {
                        this._startLogging();
                        Windows.Storage.ApplicationData.current.localSettings.values["FileLoggingSession_LoggingEnabled"] = true;
                        this._dispatchStatusChanged({ type: "LoggingEnabledDisabled", enabled: true });
                        this._setBusyStatus(false);
                        return new WinJS.Promise(function (completionFunction, errorFunction, progressFunction) {
                            completionFunction(true);
                        });
                    }
                } catch (e) {
                    this._setBusyStatus(false);
                    throw e;
                }
            },
            isLoggingEnabled: {
                get: function () {
                    return this._session !== null;
                }
            },
            _session: null,
            _channel: null,
            _isChannelEnabled: false,
            _channelLoggingLevel: Windows.Foundation.Diagnostics.LoggingLevel.verbose,
            _logFileGeneratedCount: 0,
            _isPreparingForSuspending: false,
            _isBusy: false,
            _setBusyStatus: function (busy) {
                if (this._isBusy === busy) {
                    return;
                }
                this._isBusy = busy;
                if (busy) {
                    this._dispatchStatusChanged({ type: "BusyStatusChanged" });
                }
            },
            getBusyStatus: function () {
                return this._isBusy;
            },            
            _onAppSuspension: function (suspendingEventArgs) {

                var scenario = FileLoggingSessionScenario.instance;

                // Get a deferral before performing any async operations
                // to avoid suspension prior to FileLoggingSessionScenario completing 
                // PrepareToSuspendAsync().
                var deferral = suspendingEventArgs.suspendingOperation.getDeferral();
                scenario._prepareToSuspendAsync().then(function () {
                    // From FileLoggingSessionScenario's perspective, it's now okay to 
                    // suspend, so release the deferral. 
                    deferral.complete();
                });
            },
            _onAppResume: function (resumingEventArgs) {
                var scenario = FileLoggingSessionScenario.instance;
                scenario.resumeLoggingIfApplicable();
            },
            _startLogging: function () {

                if (this._session === null) {
                    this._session = new Windows.Foundation.Diagnostics.FileLoggingSession("FileLoggingSession_Session");
                    this._session.addEventListener("logfilegenerated", this._onLogFileGenerated);
                }

                // This sample adds the channel at level "warning" to 
                // demonstrated how messages logged at more verbose levels
                // are ignored by the session. 
                this._session.addLoggingChannel(this._channel, Windows.Foundation.Diagnostics.LoggingLevel.warning);
            },
            _onLoggingEnabled: function (args) {

                // This method is called when the channel is informing us of channel-related state changes.
                // Save new channel state. These values can be used for advanced logging scenarios where, 
                // for example, it's desired to skip blocks of logging code if the channel is not being
                // consumed by any sessions. 

                var scenario = FileLoggingSessionScenario.instance;
                scenario._isChannelEnabled = args.target.enabled;
                scenario._channelLoggingLevel = args.target.level;
            },
            _onLogFileGenerated: function (args) {

                // When the log file gets large, the system closes it and starts a new one.
                // The LogFileGenerated event is raised to give the app a chance to move
                // the recently-generated log file to a safe place. When the handler returns,
                // the FileLoggingSession may reuse the file name for a new log file.
                // This function moves the log file to an app-defined location.

                var data = { scenario: FileLoggingSessionScenario.instance, finalLogFileOfSession: args.file, ourAppLogFolder: null, newLogFileName: null };
                data.scenario._logFileGeneratedCount++;

                Windows.Storage.ApplicationData.current.localFolder.createFolderAsync(
                    "FileLoggingSession_LogFiles",
                    Windows.Storage.CreationCollisionOption.openIfExists)
                .then(function (ourAppLogFolder) {
                    data.ourAppLogFolder = ourAppLogFolder;
                    data.newLogFileName = "Log-" + getTimeStamp() + ".etl";
                    return data.finalLogFileOfSession.moveAsync(data.ourAppLogFolder, data.newLogFileName, Windows.Storage.NameCollisionOption.failIfExists);
                })
                .then(function () {
                    var fullPath = data.ourAppLogFolder.path + "\\" + data.newLogFileName;
                    if (!data.scenario._isPreparingForSuspending) {
                        data.scenario._dispatchStatusChanged({ type: "LogFileGenerated", logFilePath: fullPath });
                    }
                });
            },
            resumeLoggingIfApplicable: function () {

                // If logging is already enabled, nothing to do.
                if (this.isLoggingEnabled) {
                    return true;
                }

                var loggingEnabled;
                if (Windows.Storage.ApplicationData.current.localSettings.values.hasKey("FileLoggingSession_LoggingEnabled")) {
                    loggingEnabled = Windows.Storage.ApplicationData.current.localSettings.values.hasKey("FileLoggingSession_LoggingEnabled");
                } else {
                    Windows.Storage.ApplicationData.current.localSettings.values["FileLoggingSession_LoggingEnabled"] = true;
                    loggingEnabled = true;
                }

                if (loggingEnabled) {
                    this._startLogging();
                }
                
                if (Windows.Storage.ApplicationData.current.localSettings.values.hasKey("FileLoggingSession_LogFileGeneratedBeforeSuspend")) {

                    var logFileGeneratedBeforeSuspend = Windows.Storage.ApplicationData.current.localSettings.values["FileLoggingSession_LogFileGeneratedBeforeSuspend"];
                    if (logFileGeneratedBeforeSuspend !== null) {
                        this._dispatchStatusChanged( { type: "LogFileGeneratedAtSuspend", logFilePath: logFileGeneratedBeforeSuspend });
                    }
                    Windows.Storage.ApplicationData.current.localSettings.values["FileLoggingSession_LogFileGeneratedBeforeSuspend"] = null;
                }
            },
            _closeSessionSaveFinalLogFile: function () {
                /// <summary>Call the CloseAndSaveToFileAsync() method of a FileLoggingSession, and move the log file to this app's log file repository.</summary>
                /// <returns type="WinJS.Promise">A Promise to return the final path of the log file, or null if there was no final log file.</returns>
                
                var data = { scenario: this, finalLogFileOfSession: null, ourAppLogFolder: null, newLogFileName: null };

                return this._session.closeAndSaveToFileAsync().then(function (finalLogFileOfSession) {

                    // FileLoggingSession.closeAndSaveToFileAsync can return null if nothing
                    // has been logged to the session since the last log file was saved.
                    if (finalLogFileOfSession !== null) {

                        data.finalLogFileOfSession = finalLogFileOfSession;
                        return Windows.Storage.ApplicationData.current.localFolder.createFolderAsync(
                                   "FileLoggingSession_LogFiles",
                                   Windows.Storage.CreationCollisionOption.openIfExists)
                        .then(function (ourAppLogFolder) {
                            data.ourAppLogFolder = ourAppLogFolder;
                            data.newLogFileName = "Log-" + getTimeStamp() + ".etl";
                            return data.finalLogFileOfSession.moveAsync(data.ourAppLogFolder, data.newLogFileName, Windows.Storage.NameCollisionOption.failIfExists);
                        })
                        .then(function () {
                            return data.ourAppLogFolder.path + "\\" + data.newLogFileName;
                        });
                    }

                    // There was no log file, so simply return the 'null' path to indicate there was no log file.
                    return new WinJS.Promise(function (completionFunction, errorFunction, progressFunction) {
                        completionFunction(null);
                    });
                });
            },
            _prepareToSuspendAsync: function () {
                /// <summary>Save any final log file and save app state before suspending.</summary>
                /// <returns type="WinJS.Promise">A Promise.</returns>
                if (this._session !== null) {
                    this._isPreparingForSuspending = true;
                    var data = { scenario: this };
                    return this._closeSessionSaveFinalLogFile().then(function (finalLogFilePath) {
                        data.scenario._session = null;
                        Windows.Storage.ApplicationData.current.localSettings.values["FileLoggingSession_LoggingEnabled"] = true;
                        Windows.Storage.ApplicationData.current.localSettings.values["FileLoggingSession_LogFileGeneratedBeforeSuspend"] = finalLogFilePath;
                    }).then(function () {
                        data.scenario._isPreparingForSuspending = false;
                    });
                } else {
                    Windows.Storage.ApplicationData.current.localSettings.values["FileLoggingSession_LoggingEnabled"] = false;
                    Windows.Storage.ApplicationData.current.localSettings.values["FileLoggingSession_LogFileGeneratedBeforeSuspend"] = null;
                    return new WinJS.Promise(function (completionFunction, errorFunction, progressFunction) {
                        completionFunction(); 
                    });
                }
            },
            _dispatchStatusChanged: function (statusArgs) {
                this.dispatchEvent("statusChanged", statusArgs);
            }
        },
        {
            // FileLoggingSessionScenario static members:

            _loggingScenarioSingleton: null,
            instance: {
                get: function () {
                    if (!this._loggingScenarioSingleton) {
                        this._loggingScenarioSingleton = new FileLoggingSessionScenario();
                    }
                    return this._loggingScenarioSingleton;
                }
            }
        });
    
    WinJS.Class.mix(FileLoggingSessionScenario, WinJS.Utilities.eventMixin);

    // Export public methods & controls
    WinJS.Namespace.define("FileLoggingSessionScenario", {
        FileLoggingSessionScenario: FileLoggingSessionScenario
    });

})();
